File: rspace++/src/rspace/history/instances/radix_history.rs
// See rspace/src/main/scala/coop/rchain/rspace/history/instances/RadixHistory.
// scala

use std::collections::HashSet;
use std::sync::Arc;

use shared::rust::ByteVector;
use shared::rust::store::key_value_store::KeyValueStore;

use crate::rspace::errors::HistoryError;
use crate::rspace::hashing::blake2b256_hash::Blake2b256Hash;
use crate::rspace::history::history::History;
use crate::rspace::history::history_action::{HistoryAction, HistoryActionTrait};
use crate::rspace::history::radix_tree::{Node, RadixTreeImpl, empty_node, hash_node};

pub struct RadixHistory {
    root_hash: Blake2b256Hash,
    root_node: Node,
    imple: RadixTreeImpl,
    store: Arc<dyn KeyValueStore>,
}

impl RadixHistory {
    pub fn create(
        root: Blake2b256Hash,
        store: Arc<dyn KeyValueStore>,
    ) -> Result<RadixHistory, HistoryError> {
        let imple = RadixTreeImpl::new(store.clone());
        let node = imple.load_node(root.bytes(), Some(true))?;

        Ok(RadixHistory {
            root_hash: root,
            root_node: node,
            imple,
            store,
        })
    }

    pub fn create_store(store: Arc<dyn KeyValueStore>) -> Arc<dyn KeyValueStore> { store }

    pub fn empty_root_node_hash() -> Blake2b256Hash {
        let node_hash_bytes = hash_node(&empty_node()).0;
        let node_hash = Blake2b256Hash::from_bytes(node_hash_bytes);
        node_hash
    }

    fn has_no_duplicates(&self, actions: &Vec<HistoryAction>) -> bool {
        let keys: HashSet<_> = actions.iter().map(|action| action.key()).collect();
        keys.len() == actions.len()
    }
}

impl History for RadixHistory {
    fn read(&self, key: ByteVector) -> Result<Option<ByteVector>, HistoryError> {
        let read_result = self.imple.read(&self.root_node, key.as_slice())?;
        Ok(read_result)
    }

    fn process(&self, actions: Vec<HistoryAction>) -> Result<Box<dyn History>, HistoryError> {
        if !self.has_no_duplicates(&actions) {
            return Err(HistoryError::ActionError(
                "Cannot process duplicate actions on one key.".to_string(),
            ));
        }

        let new_root_node_opt = self.imple.make_actions(&self.root_node, actions)?;

        match new_root_node_opt {
            Some(new_root_node) => {
                let node_hash_bytes = self.imple.save_node(new_root_node.clone());
                let root_hash = Blake2b256Hash::from_bytes(node_hash_bytes);
                // Avoid cloning RadixTreeImpl caches into each checkpointed history instance.
                // A fresh tree backed by the same store preserves correctness and reduces
                // allocator pressure from DashMap clone paths.
                let new_imple = RadixTreeImpl::new(self.store.clone());
                let new_history = RadixHistory {
                    root_hash,
                    root_node: new_root_node,
                    imple: new_imple,
                    store: self.store.clone(),
                };
                self.imple.commit()?;

                self.imple.clear_write_cache();
                self.imple.clear_read_cache();

                Ok(Box::new(new_history))
            }
            None => {
                Ok(Box::new(RadixHistory {
                    root_hash: self.root_hash.clone(),
                    root_node: self.root_node.clone(),
                    imple: RadixTreeImpl::new(self.store.clone()),
                    store: self.store.clone(),
                }))
            }
        }
    }

    fn root(&self) -> Blake2b256Hash { self.root_hash.clone() }

    fn reset(&self, root: &Blake2b256Hash) -> Result<Box<dyn History>, HistoryError> {
        let imple = RadixTreeImpl::new(self.store.clone());
        let node = imple.load_node(root.bytes(), Some(true))?;

        Ok(Box::new(RadixHistory {
            root_hash: root.clone(),
            root_node: node,
            imple,
            store: self.store.clone(),
        }))
    }
}

File: rspace/src/main/scala/coop/rchain/rspace/history/instances/RadixHistory.scala
package coop.rchain.rspace.history.instances

import cats.Parallel
import cats.effect.Sync
import cats.syntax.all._
import coop.rchain.rspace.hashing.Blake2b256Hash
import coop.rchain.rspace.history.RadixTree._
import coop.rchain.rspace.history._
import coop.rchain.shared.syntax.sharedSyntaxKeyValueStore
import coop.rchain.store.{KeyValueStore, KeyValueTypedStore}
import scodec.bits.ByteVector

/**
  * History implementation with radix tree
  */
object RadixHistory {
  val emptyRootHash: Blake2b256Hash = Blake2b256Hash.fromByteVector(hashNode(emptyNode)._1)

  def apply[F[_]: Sync: Parallel](
      root: Blake2b256Hash,
      store: KeyValueTypedStore[F, ByteVector, ByteVector]
  ): F[RadixHistory[F]] =
    for {
      impl <- Sync[F].delay(new RadixTreeImpl[F](store))
      node <- impl.loadNode(root.bytes, noAssert = true)
      // _ = println(
      //   "\nRadix Tree after Radix History save_node: "
      // )
      // _ <- impl
      //       .printTree(node, "", false)
    } yield RadixHistory(root, node, impl, store)

  def createStore[F[_]: Sync](
      store: KeyValueStore[F]
  ): KeyValueTypedStore[F, ByteVector, ByteVector] =
    store.toTypedStore(scodec.codecs.bytes, scodec.codecs.bytes)
}

final case class RadixHistory[F[_]: Sync: Parallel](
    rootHash: Blake2b256Hash,
    rootNode: Node,
    impl: RadixTreeImpl[F],
    store: KeyValueTypedStore[F, ByteVector, ByteVector]
) extends History[F] {
  override type HistoryF = History[F]

  override def root: Blake2b256Hash = rootHash

  override def reset(root: Blake2b256Hash): F[History[F]] =
    for {
      impl <- Sync[F].delay(new RadixTreeImpl[F](store))
      node <- impl.loadNode(root.bytes, noAssert = true)
      // _    = println("\nrootNode in reset: " + node)
    } yield this.copy(root, node, impl, store)

  override def read(key: ByteVector): F[Option[ByteVector]] =
    impl.read(rootNode, key)

  override def process(actions: List[HistoryAction]): F[History[F]] =
    for {

      /** TODO: To improve time, it is possible to implement this check into the [[RadixTreeImpl.makeActions()]]. */
      _ <- new RuntimeException("Cannot process duplicate actions on one key.").raiseError
            .unlessA(hasNoDuplicates(actions))

      // _ = println("\nhit process")

      // _              = println("\nroot_node (curr_node) in process method: " + rootNode)
      newRootNodeOpt <- impl.makeActions(rootNode, actions)
      // _              = println("\nnewRootNodeOpt: " + newRootNodeOpt)
      newHistoryOpt <- newRootNodeOpt.traverse { newRootNode =>
                        // println("\nnewRootNode in process: " + newRootNode)
                        // val hash = impl.saveNode(newRootNode)
                        // // println("\nRadix Tree after Radix History save_node: ")
                        // for {
                        //   // _                <- impl.printTree(newRootNode, "", false)
                        //   blakeHash        = Blake2b256Hash.fromByteVector(hash)
                        //   newHistory       = this.copy(blakeHash, newRootNode, impl, store)
                        //   committedHistory <- impl.commit.as(newHistory)
                        //   // store_map        <- store.toMap
                        //   // _                = println("\nstore after commit: " + store_map)
                        // } yield committedHistory

                        val hash      = impl.saveNode(newRootNode)
                        val blakeHash = Blake2b256Hash.fromByteVector(hash)
                        // println("\nnewRootNode hash: " + blakeHash.bytes)
                        val newHistory = this.copy(blakeHash, newRootNode, impl, store)
                        impl.commit.as(newHistory)
                      }
      _ = impl.clearWriteCache()
      _ = impl.clearReadCache()
      // _ = println("\nPrinting Radix Tree after process: ")
      // _ <- impl.printTree(this.rootNode, "", false)
    } yield newHistoryOpt.getOrElse(this)

  private def hasNoDuplicates(actions: List[HistoryAction]) =
    actions.map(_.key).distinct.size == actions.size
}
